src/control/system/database.go

```diff
@@ -0,0 +1,673 @@
+		if err != nil {
+			return
+		}
+
+		for _, ia := range ifaceAddrs {
+			if in, ok := ia.(*net.IPNet); ok {
+				localAddrs = append(localAddrs, in.IP)
+			}
+		}
+	} else {
+		localAddrs = append(localAddrs, ctrlAddr.IP)
+	}
+
+	for idx, candidate := range repAddrs {
+		if candidate.Port != ctrlAddr.Port {
+			continue
+		}
+		for _, localAddr := range localAddrs {
+			if candidate.IP.Equal(localAddr) {
+				repAddr = candidate
+				if idx == 0 {
+					isBootStrap = true
+				}
+				return
+			}
+		}
+	}
+
+	return
+}
+
+// ReplicaAddr returns the system's replica address if
+// the system is configured as a MS replica.
+func (db *Database) ReplicaAddr() (*net.TCPAddr, error) {
+	if !db.isReplica() {
+		return nil, &ErrNotReplica{db.cfg.Replicas}
+	}
+	return db.replicaAddr, nil
+}
+
+func (db *Database) isReplica() bool {
+	return db.replicaAddr != nil
+}
+
+func (db *Database) checkLeader() error {
+	if !db.isReplica() {
+		return &ErrNotReplica{db.cfg.Replicas}
+	}
+	if db.raft.State() != raft.Leader {
+		return &ErrNotLeader{
+			LeaderHint: string(db.raft.Leader()),
+			Replicas:   db.cfg.Replicas,
+		}
+	}
+	return nil
+}
+
+// IsLeader returns a boolean indicating whether or not this
+// system thinks that is a) a replica and b) the current leader.
+func (db *Database) IsLeader() bool {
+	return db.checkLeader() == nil
+}
+
+// OnLeadershipGained registers callbacks to be run when this instance
+// gains the leadership role.
+func (db *Database) OnLeadershipGained(fns ...onLeadershipGainedFn) {
+	db.onLeadershipGained = append(db.onLeadershipGained, fns...)
+}
+
+// OnLeadershipLost registers callbacks to be run when this instance
+// loses the leadership role.
+func (db *Database) OnLeadershipLost(fns ...onLeadershipLostFn) {
+	db.onLeadershipLost = append(db.onLeadershipLost, fns...)
+}
+
+// Start checks to see if the system is configured as a MS replica. If
+// not, it returns early without an error. If it is, the persistent storage
+// is initialized if necessary, and the replica is started to begin the
+// process of choosing a MS leader.
+func (db *Database) Start(ctx context.Context, ctrlAddr *net.TCPAddr) error {
+	var isBootStrap, needsBootStrap bool
+	var err error
+
+	db.replicaAddr, isBootStrap, err = db.checkReplica(ctrlAddr)
+	if err != nil {
+		return err
+	}
+	db.log.Debugf("system db start: isReplica: %t, isBootStrap: %t", db.isReplica(), isBootStrap)
+
+	// If we're not a replica, exit early.
+	if !db.isReplica() {
+		return nil
+	}
+
+	if _, err := os.Stat(db.cfg.RaftDir); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrapf(err, "can't Stat() %s", db.cfg.RaftDir)
+		}
+		needsBootStrap = true
+		if err := os.Mkdir(db.cfg.RaftDir, 0700); err != nil {
+			return errors.Wrapf(err, "failed to Mkdir() %s", db.cfg.RaftDir)
+		}
+	}
+
+	rc := raft.DefaultConfig()
+	rc.Logger = newCompatLogger(db.log)
+	rc.SnapshotThreshold = 16 // arbitrarily low to exercise snapshots
+	//rc.SnapshotInterval = 5 * time.Second
```

Comment:
remove or add TODO+?

Reply:
There's more tuning work required here, particularly as we add support for multiple replicas.